Add proposal for enhanced enums in C# (#9713)
dotnet/csharplang: meetings/working-groups/discriminated-unions/extended-enums.md

```csharp
// ✓ Valid - constant enum with string base
enum Status : string { Active = "A", Inactive = "I" }

// ✓ Valid - shape enum with data
enum class Result { Ok(int value), Error(string msg) }

// ✗ Invalid - cannot mix constants and shapes
enum Bad { A = 1, B(string x) }

// ✗ Invalid - shape enums cannot have base types
enum struct Bad : int { A, B }
```

For the complete formal grammar, see [Appendix A: Grammar Changes](#appendix-a-grammar-changes).

### Constant Value Enums

Enhanced constant enums support any primitive type with compile-time constants:

```csharp
enum Priority : string 
{
    Low = "low",
    Medium = "medium", 
    High = "high"
}
```

These compile to `System.Enum` subclasses with the appropriate `value__` backing field. Non-integral constant enums require explicit values for each member.

### Shape Enums: Discriminated Unions Made Elegant

Shape enums combine type unions with convenient integrated syntax:

```csharp
enum FileOperation
{
    Open(string path),
    Close,
    Read(byte[] buffer, int offset, int count),
    Write(byte[] buffer)
}
```

#### Reference Type and Value Type

> 

Author: ```suggestion

```csharp
enum class WebResponse
{
    Success(string content),
    Error(int statusCode, string message),
    Timeout
}

enum struct Option<T>
{
    None,
    Some(T value)
}
```

```csharp
enum class Result
{
    Success(string value),
    Failure(int code)
}

// Translates to:
public union Result
{
    Success,
    Failure;
    
    public sealed record class Success(string value);
    public sealed record class Failure(int code);
}
```

Singleton cases generate types with shared instances:

```csharp
enum class State { Ready, Processing, Complete }

// Translates to:
public union State
{
    Ready, Processing, Complete;
    
    public sealed class Ready 
    {
        public static readonly Ready Instance = new();
        private Ready() { }
    }
    // Similar for Processing and Complete
}
```

### `enum struct` Translation

```csharp
enum struct Option<T>
{
    None,
    Some(T value)
}

// Conceptually translates to:
public struct Option<T> : IUnion
{
    public readonly struct None { }
    public readonly record struct Some(T value);

    // Optimized layout: discriminator + space for largest case
    private byte _discriminant;
    private T _value;
    
    object? IUnion.Value => _discriminant switch
    {
        1 => new None(),
        2 => new Some(_value),
        _ => null
    };
    
    // Non-boxing access pattern

>     public bool TryGetValue(out None value) { ... }
>     public bool TryGetValue(out Some value) { ... }

Author: ```suggestion
    public bool TryGetValue(out None value)
    {
        value = default;
        return _discriminant == 1;
    }

    public bool TryGetValue(out Some value)
    {
        if (_discriminant == 2)
        {
            value = new Some(_value);
            return true;
        }
        value = default!;
        return false;
    }
```

// ✓ Valid - shape enum with data
enum class Result { Ok(int value), Error(string msg) }

// ✗ Invalid - cannot mix constants and shapes
enum Bad { A = 1, B(string x) }

// ✗ Invalid - shape enums cannot have base types
enum struct Bad : int { A, B }
```

For the complete formal grammar, see [Appendix A: Grammar Changes](#appendix-a-grammar-changes).

### Constant Value Enums

Enhanced constant enums support any primitive type with compile-time constants:

```csharp
enum Priority : string 
{
    Low = "low",
    Medium = "medium", 
    High = "high"
}
```

These compile to `System.Enum` subclasses with the appropriate `value__` backing field. Non-integral constant enums require explicit values for each member.

### Shape Enums: Discriminated Unions Made Elegant

Shape enums combine type unions with convenient integrated syntax:

```csharp
enum FileOperation
{
    Open(string path),
    Close,
    Read(byte[] buffer, int offset, int count),
    Write(byte[] buffer)
}
```

#### Reference Type and Value Type

```csharp
enum class WebResponse
{
    Success(string content),
    Error(int statusCode, string message),
    Timeout
}

enum struct Option<T>
{
    None,
    Some(T value)
}

> **`enum class`** creates discriminated unions with reference type cases:

Author: ````suggestion
```

**`enum class`** creates discriminated unions with reference type cases:
````

enum struct Option<T>
{
    None,
    Some(T value)
}
```

**`enum class`** creates discriminated unions with reference type cases:
- Cheap to pass around (pointer-sized)
- No struct tearing risk
- Natural null representation

**`enum struct`** creates discriminated unions with optimized value-type storage:
- No heap allocation
- Better cache locality
- Reduced GC pressure

#### Members and Methods

Enums can contain members just like unions:

```csharp
enum class Result<T>
{
    Success(T value),
    Error(string message);
    
    public bool IsSuccess => this switch 
    {
        Success(_) => true,
        _ => false
    };
    
    public T GetValueOrDefault(T defaultValue) => this switch
    {
        Success(var value) => value,
        _ => defaultValue
    };
}
```

Members are restricted to:
- Instance methods, properties, indexers and events (no additional state)
- Static members
- Nested types

## 5. Translation Strategy

Shape enums translate directly to unions—generating case types as nested types and creating a union that combines them.

### `enum class` Translation

```csharp
enum class Result
{
    Success(string value),
    Failure(int code)
}

// Translates to:

> public union Result(Sucess, Failure)

Reviewer: ```suggestion
public union Result(Success, Failure)
```

# Discriminated Unions and Enhanced Enums for C#

> This proposal extends C#'s union capabilities by introducing enhanced enums as algebraic sum types. While [type unions](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md) solve the problem of values that can be one of several existing types, enhanced enums provide rich, exhaustive case-based types with associated data, building on the familiar enum keyword.

Author: ```suggestion
This proposal extends C#'s union capabilities by introducing enhanced enums as [algebraic sum types](https://en.wikipedia.org/wiki/Sum_type). While [type unions](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md) solve the problem of values that can be one of several existing types, enhanced enums provide rich, exhaustive case-based types with associated data, building on the familiar enum keyword.
```

> This proposal extends C#'s union capabilities by introducing enhanced enums as algebraic data types. While [type unions](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md) solve the problem of values that can be one of several existing types, enhanced enums provide rich, exhaustive case-based types with associated data, building on the familiar enum keyword.

Author: ```suggestion
This proposal extends C#'s union capabilities by introducing enhanced enums as algebraic sum types. While [type unions](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md) solve the problem of values that can be one of several existing types, enhanced enums provide rich, exhaustive case-based types with associated data, building on the familiar enum keyword.
```

## 1. Overview

### Two Complementary Features

> C# will gain two separate features for different modeling needs: type unions (already specified) and enhanced enums (this proposal). These features work together but solve distinct problems in the type system.

Author: ```suggestion
C# will gain two separate features for different modeling needs: [type unions](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md) and enhanced enums (this proposal). These features work together but solve distinct problems in the type system.
```

- Exhaustiveness checking in switch expressions
- Enhanced nullability tracking

### Relationship to This Proposal

This proposal leaves type unions unchanged. Enhanced enums are built independently, though both features share conceptual ground in making C#'s type system more expressive. Where unions excel at "or" relationships between types, enhanced enums excel at modeling variants within a single type.

## 4. Enhanced Enums

### Design Principles

Enhanced enums follow these core principles:

- **Progressive enhancement**: Simple enums stay simple; complexity is opt-in
- **Exhaustiveness**: The compiler knows all possible cases.  See [Closed Enums](https://github.com/dotnet/csharplang/blob/main/proposals/closed-enums.md) for more details.
- **Data carrying**: Each case can carry along its own constituent data in a safe and strongly typed manner.
- **Familiar syntax**: Builds on existing enum concepts

### Syntax Extensions

Enhanced enums extend the traditional enum syntax in three orthogonal ways:

#### Extended Base Types

Traditional enums only support integral types. Enhanced enums support any constant-bearing type:

```csharp
enum Traditional : int { A = 1, B = 2 }

> enum Extended : string { Active = "active", Inactive = "inactive" }

Author: ```suggestion
enum Priority : string { Low = "low", Medium = "medium", High = "high" }
```

enum TranscendentalConstants : double { Pi = 3.14159, E = 2.71828 }
```

#### Shape Declarations

A shape enum (ADT) is created by EITHER:
- Adding `class` or `struct` after `enum`, OR (inclusive)  
- Having a parameter list on any enum member

```csharp
enum class Result { Success, Failure }  // shape enum via 'class' keyword
enum struct Result { Success, Failure } // shape enum via 'struct' keyword
enum Result { Success(), Failure() }    // shape enum via parameter lists
```

When created via parameter lists alone, it defaults to `enum class` (reference type):

```csharp
enum Result { Ok(int value), Error }    // implicitly 'enum class'
// equivalent to:
enum class Result { Ok(int value), Error }

> ```

Author: ```1suggestion
```

Shape declaration members have `record`-like semantics around concepts like equality.
````

#### Data-Carrying Cases

Shape enum members can have parameter lists to carry data:

```csharp
enum Result
{
    Success(string id),
    Failure(int code, string message)
}
```

#### Combination Rules

- **Constant enums**: Can use extended base types but NOT have parameter lists.
- **Shape enums**: Can have parameter lists but NOT specify a base type
- **Mixing cases**: Cannot mix constant values and parameterized cases in the same enum

```csharp
// ✓ Valid - constant enum with string base
enum Status : string { Active = "A", Inactive = "I" }

// ✓ Valid - shape enum with data
enum class Result { Ok(int value), Error(string msg) }

// ✗ Invalid - cannot mix constants and shapes
enum Bad { A = 1, B(string x) }

// ✗ Invalid - shape enums cannot have base types
enum struct Bad : int { A, B }
```

For the complete formal grammar specification, see [Appendix A: Grammar Changes](#appendix-a-grammar-changes).

### Constant Value Enums

Enhanced constant enums extend traditional enums to support any primitive type that can have compile-time constants:

```csharp
enum Priority : string 
{
    Low = "low",
    Medium = "medium", 
    High = "high"
}

> enum MathConstants : double

Author: ```suggestion
enum IrrationalConstants : double
```

- Subclass `System.Enum` with appropriate backing field
- Metadata preserves enum semantics for reflection
- Compatible with existing enum APIs

### Shape Enums
- **`enum class`**: Abstract class hierarchy with sealed nested classes
- **`enum struct`**: Struct with discriminant and union-style storage
- Custom attributes mark these as compiler-generated enhanced enums

### Interop Considerations

Enhanced enums maintain compatibility with:
- Existing `System.Enum` APIs where applicable
- Reflection-based frameworks
- Debugger visualization
- Binary serialization (with caveats for shape enums)

### 10. Open Questions

Several design decisions remain open:

1. Can users reference the generated nested types directly, or should they remain compiler-only?
2. Should enhanced enums support `partial` for source generators?
3. What should `default(EnumType)` produce for shape enums?
4. How should enhanced enums interact with System.Text.Json and other serializers?
5. Enums *could* allow for state, outside of the individual shape cases.  There is a clear place to store these in both the `enum class` and `enum struct` layouts.  Should we allow this? Or could it be too confusing?

> 6. Enums *could* allow for constructors, though they would likely need to defer to an existing case.  Should we allow this?  Similarly, should individual cases allow for multiple constructors?  Perhaps that is better by allowing cases to have their own record-like bodies.

Author: ```suggestion
6. Enums *could* allow for constructors, though they would likely need to defer to an existing case.  Should we allow this?  Similarly, should individual cases allow for multiple constructors?  Perhaps that is better by allowing cases to have their own record-like bodies.
7. No syntax has been presented for getting instances of data-carrying enum-members.  `new OrderStatus.Processing(...)` seems heavyweight, esp. compared to `OrderState.Pending`.  Perhaps we keep construction of data-carrying values simple, and just include the argument list, without the need for `new`.
```

### Building on Familiar Concepts

By extending the existing `enum` keyword rather than introducing entirely new syntax, enhanced enums provide progressive disclosure. Simple enums remain simple, while advanced scenarios become possible without abandoning familiar patterns.

## 3. Type Unions (Brief Overview)

### Core Concepts

Type unions are fully specified in the [unions proposal](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md#summary). They provide:

- Implicit conversions from case types to the union type
- Pattern matching that unwraps union contents
- Exhaustiveness checking in switch expressions
- Enhanced nullability tracking

### Relationship to This Proposal

This proposal leaves type unions unchanged. Enhanced enums are built independently, though both features share conceptual ground in making C#'s type system more expressive. Where unions excel at "or" relationships between types, enhanced enums excel at modeling variants within a single type.

## 4. Enhanced Enums

### Design Principles

Enhanced enums follow these core principles:

- **Progressive enhancement**: Simple enums stay simple; complexity is opt-in

> - **Exhaustiveness**: The compiler knows all possible cases

Author: ```suggestion
- **Exhaustiveness**: The compiler knows all possible cases.  See [Closed Enums](https://github.com/dotnet/csharplang/blob/main/proposals/closed-enums.md) for more details.
```

> - **Exhaustiveness**: The compiler knows all possible cases. Traditional "open" enums allow any value of the underlying type to be cast to the enum, even values not explicitly declared. "Closed" enums (see [Closed Enums proposal](https://github.com/dotnet/csharplang/blob/main/proposals/closed-enums.md)) restrict values to only those declared, enabling true exhaustiveness. **Shape enums** can be `open` or `closed`, just like constant enums.  An open enum means that consumers should expect more shapes will arise in the future, and thus all processing code must be explicitly resilient to that.  A `closed shape enum` states that all cases are known and unchanging.  Consumption code that handles all declared cases is correct, and will not receive values outside of that set in the future.

Author: ```suggestion
- **Exhaustiveness**: The compiler tracks all declared cases. Both constant and shape enums can be open or closed (see [Closed Enums proposal](https://github.com/dotnet/csharplang/blob/main/proposals/closed-enums.md)). Open enums signal that the enum author may add new cases in future versions—consumers must handle unknown cases defensively (e.g., with a default branch). Closed enums guarantee their case set is complete and will never change—the compiler ensures exhaustive matching without requiring a default case. For constant enums, "open" means values outside the declared set can be cast to the enum type. For shape enums, "open" means the author reserves the right to add new shape cases in the future, while "closed" promises the current set of shapes is final.
```

- Metadata preserves enum semantics for reflection
- Compatible with existing enum APIs

### Shape Enums
- **`enum class`**: Abstract class hierarchy with sealed nested classes
- **`enum struct`**: Struct with discriminant and union-style storage
- Custom attributes mark these as compiler-generated enhanced enums

### Interop Considerations

Enhanced enums maintain compatibility with:
- Existing `System.Enum` APIs where applicable
- Reflection-based frameworks
- Debugger visualization
- Binary serialization (with caveats for shape enums)

### 10. Open Questions

Several design decisions remain open:

1. Can users reference the generated nested types directly, or should they remain compiler-only?
2. Should enhanced enums support `partial` for source generators?
3. What should `default(EnumType)` produce for shape enums?
4. How should enhanced enums interact with System.Text.Json and other serializers?
5. Enums *could* allow for state, outside of the individual shape cases.  There is a clear place to store these in both the `enum class` and `enum struct` layouts.  Should we allow this? Or could it be too confusing?

## Appendix A: Grammar Changes

```antlr

> enum_declaration
>     : attributes? enum_modifier* 'enum' enum_kind? identifier enum_base? enum_body ';'?
>     ;
> 
> enum_kind
>     : 'class'
>     | 'struct'
>     ;

Author: ```suggestion
enum_declaration
    : attributes? enum_modifier* 'enum' ('class' | 'struct')? identifier enum_base? enum_body ';'?
    ;
```

# Discriminated Unions and Enhanced Enums for C#

This proposal extends C#'s union capabilities by introducing enhanced enums as [algebraic sum types](https://en.wikipedia.org/wiki/Sum_type). While [type unions](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md) solve the problem of values that can be one of several existing types, enhanced enums provide rich, exhaustive case-based types with associated data, building on the familiar enum keyword.

It also aims to consolidate the majority of the design space and feedback over many years in this repository, especially from:

- [#113](https://github.com/dotnet/csharplang/issues/113) - Main champion issue for discriminated unions, 7+ years of discussion and multiple LDM meetings

<details>

>   <summary>any many more...</summary>

Author: ```suggestion
  <summary>as well as...</summary>
```

```csharp
enum Result
{
    Success(string id),
    Failure(int code, string message)
}
```

#### Combination Rules

- **Constant enums**: Can use extended base types but NOT have parameter lists.
- **Shape enums**: Can have parameter lists but NOT specify a base type
- **Mixing cases**: Cannot mix constant values and parameterized cases in the same enum

```csharp
// ✓ Valid - constant enum with string base
enum Status : string { Active = "A", Inactive = "I" }

// ✓ Valid - shape enum with data
enum class Result { Ok(int value), Error(string msg) }

// ✗ Invalid - cannot mix constants and shapes
enum Bad { A = 1, B(string x) }

// ✗ Invalid - shape enums cannot have base types
enum struct Bad : int { A, B }
```

For the complete formal grammar specification, see [Appendix A: Grammar Changes](#appendix-a-grammar-changes).

### Constant Value Enums

Enhanced constant enums extend traditional enums to support any primitive type that can have compile-time constants:

```csharp
enum Priority : string 
{
    Low = "low",
    Medium = "medium", 
    High = "high"
}

enum IrrationalConstants : double
{
    Pi = 3.14159265359,
    E = 2.71828182846,
    GoldenRatio = 1.61803398875
}
```

> These compile to subclasses of `System.Enum` with the appropriate backing field `value__` with the appropriate underlying type. Unlike integral enums, non-integral constant enums require explicit values for each member.

Author: ```suggestion
These compile to subclasses of `System.Enum` with the appropriate backing field `value__` with the appropriate underlying type. Unlike integral enums, which can assign values by use of monotonically increases integers, non-integral constant enums require explicit values for each member.
```

### Shape Enums

Shape enums are C#'s implementation of algebraic sum types, allowing each case to carry different data.

#### Basic Shape Declarations

Shape enums can mix cases with and without data:

```csharp
enum FileOperation
{
    Open(string path),
    Close,
    Read(byte[] buffer, int offset, int count),
    Write(byte[] buffer)
}
```

Each case defines a constructor (and corresponding deconstructor) pattern. Cases without parameter lists are singletons, while cases with parameters create new instances.

#### Reference vs Value Semantics

**`enum class`** creates reference-type enums, stored on the heap:

```csharp
enum class WebResponse
{
    Success(string content),
    Error(int statusCode, string message),
    Timeout
}
```

Benefits:
- Cheap to pass around (pointer-sized)
- No risk of struct tearing
- Natural null representation

**`enum struct`** creates value-type enums, optimized for stack storage:

```csharp
enum struct Option<T>
{
    None,
    Some(T value)
}
```

Benefits:
- No heap allocation
- Better cache locality
- Reduced GC pressure

#### Members and Methods

> Enhanced enums can contain members just like unions:

Author: ```suggestion
Enhanced enums can contain members just like unions.  This applies to both enchanced constant and shape enums.
```

```csharp
enum class Result<T>
{
    Success(T value),
    Error(string message);
    
    public bool IsSuccess => this switch 
    {
        Success(_) => true, // or just: Success => true,
        _ => false
    };
    
    public T GetValueOrDefault(T defaultValue) => this switch
    {
        Success(var value) => value,
        _ => defaultValue
    };
}
```

Members are restricted to:

> - Methods and properties (no additional state)

Author: ```suggestion
- Methods and properties (no additional state).  Though an open question tracks if we might want to allow additional state in a shape enum.
```

> - Static members

Author: ```suggestion
- Static members. An open question tracks if that could potentially include consructors.
```

# Discriminated Unions and Enhanced Enums for C#

This proposal extends C#'s union capabilities by introducing enhanced enums as [algebraic sum types](https://en.wikipedia.org/wiki/Sum_type). While [type unions](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md) solve the problem of values that can be one of several existing types, enhanced enums provide rich, exhaustive case-based types with associated data, building on the familiar enum keyword.

It also aims to consolidate the majority of the design space and feedback over many years in this repository, especially from:

- [#113](https://github.com/dotnet/csharplang/issues/113) - Main champion issue for discriminated unions, 7+ years of discussion and multiple LDM meetings

> as well as:

Author: ```suggestion
```

    Error(int statusCode, string message),
    Timeout
}
```

Benefits:
- Cheap to pass around (pointer-sized)
- No risk of struct tearing
- Natural null representation

**`enum struct`** creates value-type enums, optimized for stack storage:

```csharp
enum struct Option<T>
{
    None,
    Some(T value)
}
```

Benefits:
- No heap allocation
- Better cache locality
- Reduced GC pressure

Similar to evolution of `records`, these variations can ship at separate times.

#### Members and Methods

Enhanced enums can contain members just like unions.  This applies to both constant and shape enums.

```csharp
enum class Result<T>
{
    Success(T value),
    Error(string message);
    
    public bool IsSuccess => this switch 
    {
        Success(_) => true, // or just: Success => true,
        _ => false
    };
    
    public T GetValueOrDefault(T defaultValue) => this switch
    {
        Success(var value) => value,
        _ => defaultValue
    };
}
```

Members are restricted to:
- Methods, operators, properties and indexers (members that add no additional state).  Though an open question tracks if we might want to allow additional state in a shape enum.

> - Static members. An open question tracks if that could potentially include consructors.

Author: ```suggestion
- Static members. An open question tracks if that could potentially include constructors.
```

    public bool IsSome => _discriminant == 1;
    
    public T GetSome() 
    {
        if (_discriminant != 1) throw new InvalidOperationException();
        return _value;
    }
}
```

For complex cases with multiple fields of different types, the compiler employs union-like storage optimization:

```csharp
enum struct Message
{
    Text(string content),
    Binary(byte[] data, int length),
    Error(int code, string message)
}

// Uses overlapping storage for fields, minimizing struct size
```

## 7. Examples and Use Cases

### Migrating Traditional Enums

Traditional enums can be progressively enhanced:

```csharp
// Step 1: Traditional enum
enum OrderStatus { Pending = 1, Processing = 2, Shipped = 3, Delivered = 4 }

// Step 2: Add data to specific states
enum OrderStatus
{
    Pending,
    Processing(DateTime startedAt),
    Shipped(string trackingNumber),
    Delivered(DateTime deliveredAt)
}

// Step 3: Add methods for common operations
enum OrderStatus
{
    Pending,
    Processing(DateTime startedAt),
    Shipped(string trackingNumber),
    Delivered(DateTime deliveredAt);
    
    public bool IsComplete => this switch
    {
        Delivered => true,
        _ => false

>     };

Author: ```suggestion
    };
    
    // Alternatively: public bool IsComplete => this is Delivered;
```

  - [#7854](https://github.com/dotnet/csharplang/issues/7854) - Exhaustiveness checking for ADT patterns using private constructors
  - [#8942](https://github.com/dotnet/csharplang/discussions/8942) - Track subtype exhaustiveness for closed hierarchies
  - [#8926](https://github.com/dotnet/csharplang/discussions/8926) - Extensive discussion on Option<T> as canonical DU use case
  - [#7010](https://github.com/dotnet/csharplang/discussions/7010) - Union types discussion heavily featuring Option<T> and Result<T>
  - [#274](https://github.com/dotnet/csharplang/discussions/274) - Java-style class-level enums with methods and constructors
</details>

## 1. Overview

### Two Complementary Features

C# will gain two separate features for different modeling needs: [type unions](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md) and enhanced enums (this proposal). These features work together but solve distinct problems in the type system.

### At a Glance

```csharp
// Type unions - one value, multiple possible types
union Result { string, ValidationError, NetworkException }

// Enhanced enums - one type, multiple possible shapes  

> enum PaymentResult

Reviewer 2: Aren't enhanced enums a superset of type unions?

Given

```cs
union Result { string, ValidationError, NetworkException }
```

I could just represent it as

```cs
enum struct Result { Text(string value), ValidationError(ValidationError value), NetworkException(NetworkException value) }
```

and the usage of such enhanced enum would be mostly similar to the type union I think. The only difference would be slightly more verbose syntax at the declaration side (but we can always add more syntactic sugar). Hence it'd be enough to have enhanced enums and not type unions?

Author: Yes. There is an argument to be made here that unions are just a very very terse way of annotating a value to say it is one of a few types. And yes, it would be a good question to ask if that is valuable enough. 

Personally,I think it is. But the conversation is a good one:-)

Enhanced constant enums are similar to classical enums in that they are open by default, but can be potentially 'closed' (see [Closed Enums](https://github.com/dotnet/csharplang/blob/main/proposals/closed-enums.md)).  Open and closed enums with non-integral backing types behave similarly to their integral counterparts.  For example, allowing/disallowing conversions from their underlying type, and treating pattern matching as exhaustive or not depending on if all declared values were explicitly matched.

### Shape Enums

Shape enums are C#'s implementation of algebraic sum types, allowing each case to carry different data.

#### Basic Shape Declarations

Shape enums can mix cases with and without data:

```csharp
enum FileOperation
{
    Open(string path),
    Close,
    Read(byte[] buffer, int offset, int count),
    Write(byte[] buffer)
}
```

Each case defines a constructor (and corresponding deconstructor) pattern. Cases without parameter lists are singletons, while cases with parameters create new instances.

#### Reference vs Value Semantics

> **`enum class`** creates reference-type enums, stored on the heap:

Author: ```suggestion
**`enum class`** creates reference-type enums, stored according to underlying runtime choices (normally the heap):
```

2. A union declaration combining these types

```csharp
enum class Result
{
    Success(string value),
    Failure(int code)
}

// Conceptually translates to:

// Generated union
public union Result
{
    Success,
    Failure;
    
    // Generated case types
    public sealed record class Success(string value);
    public sealed record class Failure(int code);
}

Singleton cases (those without parameters) generate types with shared instances:

```csharp
enum class State { Ready, Processing, Complete }

public union State
{
    Ready, Processing, Complete;
    
    // Conceptually translates to:
    public sealed class Ready 
    {
        public static readonly Ready Instance = new();
        private Ready() { }
    }
    // Similar for Processing and Complete
}
```

### `enum struct` Translation

An `enum struct` also generates case types and a union, but the union uses an optimized storage layout as permitted by the [non-boxing union access pattern](https://raw.githubusercontent.com/dotnet/csharplang/refs/heads/main/proposals/unions.md#union-patterns):

```csharp
enum struct Option<T>
{
    None,
    Some(T value)
}

// Conceptually translates to:

> // Generated case types
> public readonly struct Option_None { }
> public readonly record struct Option_Some<T>(T value);
> 
> // Generated union with optimized storage
> public struct Option<T> : IUnion
> {
>     // Optimized layout: discriminator + space for largest case
>     private byte _discriminant;
>     private T _value;  // Space for Some's data
>     
>     // Implements IUnion.Value
>     object? IUnion.Value => _discriminant switch
>     {
>         1 => new Option_None(),
>         2 => new Option_Some<T>(_value),
>         _ => null
>     };
>     
>     // Non-boxing access pattern
>     public bool HasValue => _discriminant != 0;
>     
>     public bool TryGetValue(out Option_None value)
>     {
>         value = default;
>         return _discriminant == 1;
>     }
>     
>     public bool TryGetValue(out Option_Some<T> value)
>     {
>         if (_discriminant == 2)
>         {
>             value = new Option_Some<T>(_value);
>             return true;
>         }
>         value = default!;
>         return false;
>     }

Author: ```suggestion
// Generated union with optimized storage
public struct Option<T> : IUnion
{
    // Generated case types
    public readonly struct None { }
    public readonly record struct Some<T>(T value);

    // Optimized layout: discriminator + space for largest case
    private byte _discriminant;
    private T _value;  // Space for Some's data
    
    // Implements IUnion.Value
    object? IUnion.Value => _discriminant switch
    {
        1 => newNone(),
        2 => new Some<T>(_value),
        _ => null
    };
    
    // Non-boxing access pattern
    public bool HasValue => _discriminant != 0;
    
    public bool TryGetValue(out None value)
    {
        value = default;
        return _discriminant == 1;
    }
    
    public bool TryGetValue(out Some<T> value)
    {
        if (_discriminant == 2)
        {
            value = new Some<T>(_value);
            return true;
        }
        value = default!;
        return false;
    }
```

>     // Constructors
>     public Option(Option_None _) => _discriminant = 1;
>     public Option(Option_Some<T> some) => (_discriminant, _value) = (2, some.value);
>     
>     // Convenience factories
>     public static Option<T> None => new Option<T>(new Option_None());
>     public static Option<T> Some(T value) => new Option<T>(new Option_Some<T>(value));

Author: ```suggestion
    // Constructors
    public Option(None _) => _discriminant = 1;
    public Option(Some<T> some) => (_discriminant, _value) = (2, some.value);
    
    // Convenience factories
    public static Option<T> None => new Option<T>(new None());
    public static Option<T> Some(T value) => new Option<T>(new Some<T>(value));
```

#### Members and Methods

Enums can contain members just like unions. This applies to both constant and shape enums:

```csharp
enum class Result<T>
{
    Success(T value),
    Error(string message);
    
    public bool IsSuccess => this switch 
    {
        Success(_) => true,
        _ => false
    };
    
    public T GetValueOrDefault(T defaultValue) => this switch
    {
        Success(var value) => value,
        _ => defaultValue
    };
}
```

Members are restricted to:
- Instance methods, properties and indexers (members that add no additional state)
- Static members
- Nested types

## 5. Translation Strategy

### Shape Enum Translation Overview

Shape enums are syntactic sugar that generates:
1. Individual case types (as records or similar types)
2. A union type that combines these cases
3. Convenience members for construction and pattern matching

### `enum class` Translation

An `enum class` generates:
1. A set of record class types for each case
2. A union declaration combining these types

```csharp
enum class Result
{
    Success(string value),
    Failure(int code)
}

// Conceptually translates to:

> // Generated case types
> public sealed record class Result_Success(string value);
> public sealed record class Result_Failure(int code);
> 
> // Generated union
> public union Result(Result_Success, Result_Failure)
> {
>     // Convenience constructors/factories
>     public static Result Success(string value) => new Result(new Result_Success(value));
>     public static Result Failure(int code) => new Result(new Result_Failure(code));
> }
> ```

Author: ```suggestion

// Generated union
public union Result
{
    Success,
    Failure;
    
    // Generated case types
    public sealed record class Success(string value);
    public sealed record class Failure(int code);
}
```
